File: .skills/openclaw-skills/skills/billc8128/openclaw-scribe/scripts/setup-cron.py
#!/usr/bin/env python3
"""
Scribe setup — registers a nightly OpenClaw cron job to run scribe automatically.

Usage:
    python3 scripts/setup-cron.py

This script calls the OpenClaw Gateway API to register a cron job that runs
scribe every night at 23:30 (local timezone). The job runs in an isolated
session and writes memory/YYYY-MM-DD.md to your workspace.
"""

import json
import os
import sys
import subprocess
from pathlib import Path

GATEWAY_DEFAULT = "http://127.0.0.1:19000"
CONFIG_PATH = Path.home() / ".openclaw/openclaw.json"

CRON_JOB = {
    "name": "scribe-nightly",
    "schedule": {
        "kind": "cron",
        "expr": "30 23 * * *",
        "tz": "Asia/Shanghai",    # change to your timezone
    },
    "payload": {
        "kind": "agentTurn",
        "message": (
            "Run the scribe skill now. "
            "Execute: python3 skills/public/scribe/scripts/scribe.py — "
            "scan today's session logs from ~/.openclaw/agents/main/sessions/ "
            "and write structured memory to memory/YYYY-MM-DD.md in the workspace. "
            "Report how many messages were processed and where the file was written."
        ),
    },
    "sessionTarget": "isolated",
    "enabled": True,
}


def get_gateway_url() -> str:
    url = os.environ.get("OPENCLAW_GATEWAY_URL", "")
    if url:
        return url.rstrip("/")
    try:
        cfg = json.loads(CONFIG_PATH.read_text())
        gw = cfg.get("gateway", {})
        port = gw.get("port") or gw.get("httpPort") or 19000
        return f"http://127.0.0.1:{port}"
    except Exception:
        return GATEWAY_DEFAULT


def get_gateway_token() -> str:
    token = os.environ.get("OPENCLAW_GATEWAY_TOKEN", "")
    if token:
        return token
    try:
        cfg = json.loads(CONFIG_PATH.read_text())
        gw = cfg.get("gateway", {})
        # token may be at gateway.token or gateway.auth.token
        return gw.get("token") or gw.get("auth", {}).get("token", "")
    except Exception:
        return ""


def curl_json(method: str, url: str, token: str, data=None) -> dict:
    cmd = ["curl", "-s", "-X", method, url,
           "-H", "Content-Type: application/json",
           "-H", f"Authorization: Bearer {token}"]
    if data:
        cmd += ["-d", json.dumps(data)]
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=15)
    if result.returncode != 0:
        raise RuntimeError(f"curl failed: {result.stderr}")
    return json.loads(result.stdout)


def main():
    gateway = get_gateway_url()
    token = get_gateway_token()

    print(f"[setup] Gateway: {gateway}")

    if not token:
        print("[setup] ❌ No gateway token found.")
        print("        Set OPENCLAW_GATEWAY_TOKEN or check ~/.openclaw/openclaw.json")
        sys.exit(1)

    # Check if job already exists
    try:
        jobs = curl_json("GET", f"{gateway}/api/cron/jobs", token)
        existing = [j for j in jobs.get("jobs", []) if j.get("name") == "scribe-nightly"]
        if existing:
            print(f"[setup] ⚠️  A cron job named 'scribe-nightly' already exists (id: {existing[0].get('id')}).")
            answer = input("         Replace it? [y/N] ").strip().lower()
            if answer != "y":
                print("[setup] Aborted.")
                sys.exit(0)
            # Delete existing
            job_id = existing[0].get("id") or existing[0].get("jobId")
            curl_json("DELETE", f"{gateway}/api/cron/jobs/{job_id}", token)
            print(f"[setup] Deleted existing job {job_id}")
    except Exception as e:
        print(f"[setup] ⚠️  Could not list existing jobs: {e}")

    # Register new job
    try:
        resp = curl_json("POST", f"{gateway}/api/cron/jobs", token, CRON_JOB)
        job_id = resp.get("id") or resp.get("jobId") or resp.get("job", {}).get("id")
        print(f"[setup] ✅ Cron job registered! id={job_id}")
        print(f"         Schedule: every night at 23:30 (Asia/Shanghai)")
        print(f"         Scribe will write to: memory/YYYY-MM-DD.md")
        print()
        print("To run immediately:  python3 skills/public/scribe/scripts/scribe.py")
        print("To change timezone:  edit CRON_JOB['schedule']['tz'] in this script")
    except Exception as e:
        print(f"[setup] ❌ Failed to register cron job: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: .skills/openclaw-skills/skills/ramboxie/claw-rpg/scripts/setup-cron.mjs
#!/usr/bin/env node
/**
 * Claw RPG — 自动设置 XP 同步 Cron
 * 每日 03:00 同步 XP（通过 OpenClaw cron 系统）
 *
 * 用法：node scripts/setup-cron.mjs
 */

import { readFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { SKILL_ROOT, SCRIPTS } from './_paths.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));

// 读取 OpenClaw gateway token
function loadGatewayToken() {
  const paths = [
    join(process.env.USERPROFILE || '', '.openclaw', 'openclaw.json'),
    join(process.env.HOME || '', '.openclaw', 'openclaw.json'),
  ];
  for (const p of paths) {
    if (existsSync(p)) {
      try { return JSON.parse(readFileSync(p, 'utf8'))?.gateway?.auth?.token; } catch {}
    }
  }
  return null;
}

async function run() {
  const token = loadGatewayToken();
  if (!token) {
    console.error('❌ 未找到 OpenClaw gateway token');
    console.log('   请确认 OpenClaw 已安装并运行');
    process.exit(1);
  }

  const port = 18789;

  // 创建每日 03:00 XP 同步 cron
  const job = {
    name: 'claw-rpg-daily-xp',
    schedule: { kind: 'cron', expr: '0 3 * * *', tz: 'Asia/Shanghai' },
    payload: {
      kind: 'systemEvent',
      text: `[Claw RPG] 每日 XP 同步提醒：请运行 node ${SCRIPTS}/xp.mjs 更新今日 XP（使用 session_status 获取 token delta）`
    },
    sessionTarget: 'main',
    enabled: true,
  };

  try {
    const res = await fetch(`http://localhost:${port}/cron/jobs`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(job),
    });

    if (res.ok) {
      const data = await res.json();
      console.log('✅ Cron 已设置：每日 03:00 XP 同步提醒');
      console.log(`   Job ID: ${data.id || '已创建'}`);
    } else {
      const err = await res.text();
      // Cron API 路径可能不同，提示手动设置
      console.log('⚠️  自动设置失败，请手动在 OpenClaw 中添加以下 cron：');
      printManualInstructions();
    }
  } catch (e) {
    console.log('⚠️  无法连接 OpenClaw gateway，请手动添加 cron：');
    printManualInstructions();
  }
}

function printManualInstructions() {
  console.log(`
  在 OpenClaw 中手动添加每日 XP 同步 cron：
  
  Schedule: 每日 03:00
  Payload (systemEvent):
    "[Claw RPG] 请运行 node ${SCRIPTS}/xp.mjs 更新今日 XP"
  
  或直接在 HEARTBEAT.md 中加入：
    # Claw RPG XP 同步
    每 20 次对话，运行：node ${SCRIPTS}/xp.mjs --in <delta_in> --out <delta_out>
  `);
}

run().catch(e => { console.error('❌', e.message); });
